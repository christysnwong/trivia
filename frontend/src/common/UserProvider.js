import React, {useEffect, useState} from "react";
import UserContext from "./UserContext";

const demoUser = {
  id: 1,
  username: "testuser",
  first_name: "Test",
  last_name: "User",
  email: "test@test.com",
  stats: {
    level: 14,
    title: "Pro",
    "quizzes completed": 34,
    points: 3440,
    remainingPts: 60,
    levelPts: 300,
  },
};

const UserProvider = ({
  children,
  currUser = demoUser
}) => {
  
  const [currentUser, setCurrUser] = useState();

  return (
    // <UserContext.Provider value={{ currUser }}>
    <UserContext.Provider value={{ currUser, setCurrUser }}>
      {children}
    </UserContext.Provider>
  );
};

export default UserProvider;
