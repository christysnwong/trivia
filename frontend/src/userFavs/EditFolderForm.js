import { useContext, useEffect, useState } from "react";
import UserContext from "../common/UserContext";
import JoblyApi from "../api/api";

import Noty from "noty";

const EditFolderForm = ({
  folderId,
  folderName,
  setUserFolders,
  setEditFormIsOpen,
}) => {
  const { currUser } = useContext(UserContext);

  const [formData, setFormData] = useState({
    userId: currUser.id,
    newFolderName: folderName,
  });

  // handle changes in the form
  const handleChange = (evt) => {
    const { name, value } = evt.target;
    setFormData((formData) => ({
      ...formData,
      [name]: value,
    }));
  };

  // handle form submission and reset form
  const handleSubmit = async (evt) => {
    evt.preventDefault();

    try {
      let res = await JoblyApi.renameFolder(
        currUser.username,
        folderId,
        formData
      );

      if (res.updated) {
        
        new Noty({
          type: "success",
          layout: "topRight",
          theme: "bootstrap-v4",
          text: `Your folder is successfully renamed from ${folderName} to ${res.updated.name}.`,
          timeout: "3000",
          progressBar: true,
          closeWith: ["click"],
          killer: true,
        }).show();

        setUserFolders((folders) => [
          ...folders.filter(
            (folder) => folder.folderId !== res.updated.folderId
          ),
          { folderId: res.updated.folderId, name: res.updated.name },
        ]);

        setEditFormIsOpen(false);
      }

      setFormData({ userId: currUser.id, newFolderName: "" });
      
    } catch (e) {
      console.error(e);

      new Noty({
        type: "error",
        layout: "topRight",
        theme: "bootstrap-v4",
        text: e,
        timeout: "3000",
        progressBar: true,
        closeWith: ["click"],
        killer: true,
      }).show();
    }
  };

  // load info for existing folder in form when folder id / name changes
  useEffect(() => {
    setFormData((formData) => ({ ...formData, newFolderName: folderName }));
  }, [folderId, folderName]);

  return (
    <>
      <div className="card mb-5">
        <div className="card-body">
          <form onSubmit={handleSubmit}>
            <div className="mb-2">
              <label htmlFor="username" className="form-label">
                Edit a Folder name:{" "}
              </label>

              <input
                id="newFolderName"
                name="newFolderName"
                className="form-control"
                value={formData.newFolderName}
                onChange={handleChange}
                maxLength="20"
              />
            </div>

            <button
              type="button"
              className="btn btn-secondary mt-4 me-3"
              onClick={() => setEditFormIsOpen(false)}
              method="get"
            >
              Cancel
            </button>

            <button
              type="submit"
              className="btn btn-primary mt-4"
              method="post"
            >
              Save
            </button>
          </form>
        </div>
      </div>
    </>
  );
};

export default EditFolderForm;
