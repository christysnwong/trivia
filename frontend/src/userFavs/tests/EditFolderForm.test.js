import React from "react";
import { render } from "@testing-library/react";
import EditFolderForm from "../EditFolderForm";
import { MemoryRouter } from "react-router-dom";
import UserProvider from "../../common/UserProvider";

it("renders without crashing", function () {
  render(
    <UserProvider>
      <EditFolderForm />
    </UserProvider>
  );
});

it("matches snapshot", function () {
  const { asFragment } = render(
    <MemoryRouter>
      <UserProvider>
        <EditFolderForm />
      </UserProvider>
    </MemoryRouter>
  );
  expect(asFragment()).toMatchSnapshot();
});
