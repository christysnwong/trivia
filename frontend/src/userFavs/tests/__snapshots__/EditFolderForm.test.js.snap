// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="card mb-5"
  >
    <div
      class="card-body"
    >
      <form>
        <div
          class="mb-2"
        >
          <label
            class="form-label"
            for="username"
          >
            Edit a Folder name: 
          </label>
          <input
            class="form-control"
            id="newFolderName"
            maxlength="20"
            name="newFolderName"
            value=""
          />
        </div>
        <button
          class="btn btn-secondary mt-4 me-3"
          method="get"
          type="button"
        >
          Cancel
        </button>
        <button
          class="btn btn-primary mt-4"
          method="post"
          type="submit"
        >
          Save
        </button>
      </form>
    </div>
  </div>
</DocumentFragment>
`;
